### ISSUE TITLE
notation priorities and coercions

### ISSUE THREAD

**Issue by avigad** [leanprover/lean3#768]
I am trying to understand how priorities for notations interact with coercions. I did this in data/finset/basic:

``` lean
protected definition prio : num := num.succ std.priority.default
infix [priority finset.prio] ∪ := union
```

We also have this coercion, define in data/finset/to_set:

``` lean
definition to_set [coercion] (s : finset A) : set A := λx, x ∈ s
abbreviation ts := @to_set A
```

I get this behavior:

``` lean
import data.set data.finset
open set finset

variables (A : Type) [deceqA : decidable_eq A]
include deceqA
variables s t : finset A

set_option pp.coercions true
set_option pp.notation false
set_option pp.full_names true
check (s ∪ t : set A)
-- check result:
-- set.union (finset.ts s) (finset.ts t) : set A
```

In other words, Lean inserts the coercion first, rather than interpreting the union as finset.union and inserting the coercion afterwards.

With integers, it goes the other way:

``` lean
import data.int
open nat int

variables x y : nat

set_option pp.coercions true
set_option pp.notation false
set_option pp.full_names true
check (x + y : int)
-- check result:
-- int.of_nat (nat.add x y) : int
```

In other words, here Lean uses nat addition, and inserts the coercion afterwards.

Is this expected? In other words, am I doing something wrong, or should we expect that the interaction between coercions and notation is nondeterministic?

**Comment by leodemoura** [comment]
> Is this expected? In other words, am I doing something wrong, or should we expect that the interaction between coercions and notation is nondeterministic?

It is not expected and it is not due to nondeterministic behavior.
The process is deterministic. The only feature in the elaborator that produces nondeterministic behavior is the `par` tactic combinator. If we use `(par t1 t2)`, we are executing `t1` and `t2` in parallel, and the result may be different in each execution. However, this is not the case here, and we don't use `par` in the standard library.

The different behavior is a consequence of the fact that `int` and `real` are non-parametric types (i.e., they don't have an argument `A : Type`), but `set` and `finset` are.

Another issue is how lean is processing the notation `(t : A)` when `t` has overloaded expressions.
Right now, it creates a "choice-expression" for `t` and then enforces type `A`.
That is, we have something like `(enforce_type (choice t_1 ... t_n) A)`, where each `t_i` is a different interpretation for `t`. 
To fix this issue, I distributed `enforce_type` over `choice`. Now, lean produces
`choice (enforce_type t_1 A) ... (enforce_type t_n A))`. I think this is closer to what the user wants. We enforce the desired type when we are processing each choice. By enforcing the desired type, a coercion may be introduced.

**Comment by avigad** [comment]
Thanks. That makes sense. I think the change is good. It is not entirely clear to me priority for notation should always be "stronger" than the preference to not insert a coercion, but it seems like the right thing. In the `(x + y : int)` example, users force the coercion to be inserted late by putting a higher-priority on nat addition, and early by putting a higher-priority on int addition.
